Change asset_bundle_package_font_test to memory file system

DanTup opened the pull request:
This is to work towards being able to run the tests without `-j1` (#21113). These tests were using the real filesystem and setting/relying on fs.currentDirectory. There was a comment about this being because the memory provider didnt' support POSIX and Windows, however that seems to have changed since (and many other asset tests already do something similar to this).

DanTup committed:
Change assert_bundle_package_font_test to memory file system

This is to work towards being able to run the tests without `-j1` (#21113). These tests were using the real filesystem and setting/relying on fs.currentDirectory. There was a comment about this being because the memory provider didnt' support POSIX and Windows, however that seems to have changed since (and many other asset tests already do something similar to this).

DanTup commented:
@goderbauer This undoes some of your changes from https://github.com/flutter/flutter/pull/14246 because the memory file system now supports both POSIX+Windows so you may have opinions on this. I'm trying to fix tests to work concurrently (to speed up bots), starting with those using currentDirectory on the real filesystem.

DanTup committed:
Trim trailing whitespace

goderbauer commented on packages/flutter_tools/test/asset_bundle_package_fonts_test.dart:
```diff
@@ -89,25 +96,38 @@ $fontsSection
-  // asset bundles work correctly on Windows and Posix systems.
-  Directory tempDir;
-  Directory oldCurrentDir;
+  // TODO(dantup): This function exists in many tests, is there somewhere appropriate
```
You can just create a file in the test directory, name it appropriately (do not end it with _test) and put all the logic there. We have done that before in other places, for example: https://github.com/flutter/flutter/blob/master/packages/flutter/test/widgets/semantics_tester.dart

DanTup replied:
It was the "naming appropriately" bit I wasn't sure about. I've added a new file in `test/src/schema.dart` but it feels a bit only having two functions in it! LMK if you have a better idea - I've changed the other code that had copies of this (it's all the same - just making an empty schema because it's required to parse the pubspec) to use this one.

goderbauer reviewed (approved):
LGTM

Thanks for doing this!

DanTup committed:
Add a workaround for Windows path slash directions

Strictly this is correct, but the real FS can tolerate either path. The in-memory file system is more strict (see https://github.com/google/file.dart/issues/112).

DanTup committed:
Extract a helper for writing schema files in tests

DanTup committed:
Missed file when saving!

DanTup committed:
Remove redundant comment

DanTup commented:
The Windows failure was due to the in-memory file system being less tolerant of slash directions than the real fs. In the real fs we can give it paths in the wrong direction and they work (and this is what was happening in these tests). however the in-memory file system required them to match.

I've opened https://github.com/google/file.dart/issues/112 about making these behave the same, but for now I've added a helper that just fixes up slashes. The tests are passing locally on my Windows box now, so hopefully will go green.

Please check the changes for the Windows workaround and also the naming/location of the file I extracted schema file (I did consider just putting it in `common.dart`, but it's not *that* common).

goderbauer commented on packages/flutter_tools/test/src/schema.dart:
```diff
@@ -0,0 +1,19 @@
+import 'package:flutter_tools/src/base/file_system.dart';
+import 'package:flutter_tools/src/flutter_manifest.dart';
+
+/// Writes a schemaData, required for pubspec.yaml to be loaded.
```
This should probably have a better doc comment. What is schemaData?

DanTup replied:
I wasn't entirely sure, I just brought this comment with the code. However, looking more, it seems like the asset code validates pubspec.yamls using schemas. These methods are used in tests to write those schemas (and in most cases write an empty file that will pass any file).

I've tweaked the comments to be more descriptive.

goderbauer commented on packages/flutter_tools/test/src/schema.dart:
```diff
@@ -0,0 +1,19 @@
+}
+
+/// Writes an empty schemaData. A schema file is required for pubspec.yaml to be loaded.
+void writeBasicSchemaFile(FileSystem filesystem) {
```
Maybe this should be called writeEmptySchemaFile?

DanTup replied:
Done!

goderbauer commented on packages/flutter_tools/test/src/schema.dart:
```diff
@@ -0,0 +1,19 @@
+import 'package:file/file.dart';
```
Maybe name this pubspec_schema to be a little more specific?

DanTup replied:
Done!

DanTup commented on packages/flutter_tools/test/android/gradle_test.dart:
```diff
@@ -331,16 +330,6 @@ flutter:
-  final String schemaDir = buildSchemaDir(filesystem);
-
-  filesystem.directory(schemaDir).createSync(recursive: true);
-  filesystem.file(schemaFile).writeAsStringSync(schemaData);
```
Weird - in the old code (which I moved to be shared) we're passing a `File` into `filesystem.file()`. The `file` method takes `dynamic` so this didn't produce any errors. I noticed because I found another(!) copy of this function and it was slightly different again.

DanTup replied:
Apparently by design:

> path can be either a `String`, a `Uri`, or a `FileSystemEntity`.

I thought maybe it was re-using the instance as a performance thing, however it reads the `.path` of it and constructors another one, so I guess not.

DanTup committed:
Rename writeBasicSchema -> writeEmptySchema

DanTup committed:
Use the file we already have to write contents

DanTup committed:
Make comments more descriptive

DanTup committed:
Remove another dupe of writeSchema to use the shared one

DanTup committed:
Rename schema -> pubspec_schema

DanTup committed:
Trim whitespace

DanTup commented:
Failures were trailing whitespace in a file I'd touched. I've trimmed it and also rebased on master to get latest code. 

@goderbauer I pushed a couple of changes since your LGTM (mostly your comments, but also found another copy of this function, for ex.), lmk if you're still happy with this.

goderbauer reviewed (approved):
LGTM
